src/transformers/models/mctc/modeling_mctc.py

```diff
@@ -0,0 +1,904 @@
+        return x.reshape(*reversed(shape)).permute(*reversed(range(len(shape))))
+
+    def relativePositionEmbeddingRotate(self, data):
+        # NOTE: must re-evaluate whether this re-implementation was trly necessary
+        # or the reason why my complete re-haul worked was due to some other part
+        # of the code. Adding this and the reshape fortrain code seems very undesirable.
+
+        data = data.permute(0, 2, 3, 1)
+
+        b, d0, d1, d2 = data.shape
+
+        # data = af::join(0, data, af::constant(0.0, d1, d1, d2, d3, data.type()));
+        data = torch.cat((data, torch.zeros((b, d1, d1, d2), device=data.device)), dim=1)
+
+        # data = af::moddims(data, af::dim4((d0 + d1) * d1, 1, d2, d3));
+        # data = data.reshape()
+        data = self.reshape_fortran(data, [b, (d0 + d1) * d1, 1, d2])
+        # data = data.numpy().reshape((d0 + d1) * d1, 1, d2,order='F')
+
+        # data = data.rows(0, (d1 + d0 - 1) * d1 - 1);
+        data = data[:, : (d1 + d0 - 1) * d1]
+
+        # data = af::moddims(data, af::dim4(d0 + d1 - 1, d1, d2, d3));
+        data = self.reshape_fortran(data, [b, d0 + d1 - 1, d1, d2])
+
+        n = d0 // 2
+        data = data[:, n : n + d1].transpose(1, 2)
+
+        return data.permute(0, 3, 1, 2)
+
+    def forward(
+        self,
+        hidden_states,
+        attention_mask=None,
+        head_mask=None,
+        encoder_hidden_states=None,
+        encoder_attention_mask=None,
+        past_key_value=None,
+        output_attentions=False,
+    ):
+        mixed_query_layer = self.query(hidden_states)
+        mixed_query_layer = mixed_query_layer / math.sqrt(self.attention_head_size)
+
+        if past_key_value is not None:
+            key_layer = self.transpose_for_scores(self.key(hidden_states))
+            value_layer = self.transpose_for_scores(self.value(hidden_states))
+            key_layer = torch.cat([past_key_value[0], key_layer], dim=2)
+            value_layer = torch.cat([past_key_value[1], value_layer], dim=2)
+        else:
+            key_layer = self.transpose_for_scores(self.key(hidden_states))
+            value_layer = self.transpose_for_scores(self.value(hidden_states))
+
+        query_layer = self.transpose_for_scores(mixed_query_layer)
+
+        # Take the dot product between "query" and "key" to get the raw attention scores.
+        attention_scores = torch.matmul(query_layer, key_layer.transpose(-1, -2))
+
+        if self.position_embedding_type == "relative_key":
+            positional_embedding = self.distance_embedding.weight
+            relative_position_scores = torch.einsum(
+                "lh, bche -> bcle", positional_embedding, query_layer.transpose(2, 3)
+            )
+
+            relative_position_scores = self.relativePositionEmbeddingRotate(relative_position_scores)
+            attention_scores = attention_scores + relative_position_scores
+
+        # if attention_mask is not None:
```

Comment:
attention mask doesn't work yet?

Reply:
works now! forgot to uncomment things I commented out during debugging